src/easylink/step.py

```diff
@@ -261,6 +277,122 @@ def update_implementation_graph(
     def validate_step(self, step_config: LayeredConfigTree) -> Dict[str, List[str]]:
         if not self.name in step_config or not self.config_key in step_config[self.name]:
             return BasicStep.validate_step(self, step_config)
+        sub_config = step_config[self.name][self.config_key]
+        return CompositeStep.validate_step(self, sub_config)
+
+
+class LoopStep(CompositeStep, BasicStep):
+    """A LoopStep allows a user to loop a single step a user-configured number of times."""
+
+    # The first version of LoopStep relies on several (not-generalizable) assumptions
+    # First, that LoopStep is initialized with only one node with the same name
+    # as the LoopStep. It also assumes that there is only one input and output slot.
+    # We use a self-loop to represent how one loop feeds into the next.
+
+    def __init__(
+        self,
+        step_name: str,
+        name: str = None,
+        input_slots: List[InputSlot] = [],
+        output_slots: List[OutputSlot] = [],
+        nodes: List[Step] = [],
+        edges: List[Edge] = [],
+        # slot_mappings: Dict[str, List[SlotMapping]] = {"input": [], "output": []},
+    ) -> None:
+        super(CompositeStep, self).__init__(step_name, name, input_slots, output_slots)
+        # TODO [MIC-5135]: Make loopstep compatible with sequence of steps using step hierarchy (composite steps)
+        if len(nodes) != 1 or nodes[0].name != step_name:
+            raise NotImplementedError(
+                f"LoopStep {self.name} must be initialized with a single node with the same name."
+            )
+        if not isinstance(nodes[0], BasicStep):
+            raise NotImplementedError(
+                f"LoopStep {self.name} can currently only loop a single implementation."
+            )
+        self.unlooped_node = nodes[0]
+        for edge in edges:
+            if not edge.in_node == edge.out_node == step_name:
+                raise NotImplementedError(
+                    f"LoopStep {self.name} must be initialized with only self-loops as edges"
+                )
+        self.edges = edges
+
+    @property
+    def config_key(self):
+        return "iterate"
+
+    def update_implementation_graph(
+        self, graph: nx.MultiDiGraph, step_config: LayeredConfigTree
+    ) -> None:
+        if not self.name in step_config or not self.config_key in step_config[self.name]:
```

Comment:
This condition is difficult for me to parse. Is it when an implementation is given directly instead of an `iterate` list? Is there a way to make it more readable?

Reply:
hopefully this makes more sense now?